src/app/icd/client/CheckInHandler.h

```diff
@@ -0,0 +1,146 @@
+ *      handler
+ *
+ */
+
+#pragma once
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandSender.h>
+#include <app/OperationalSessionSetup.h>
+#include <app/icd/client/CheckInDelegate.h>
+#include <app/icd/client/DefaultICDClientStorage.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+
+namespace chip {
+namespace app {
+
+class CheckInHandler : public Messaging::ExchangeDelegate, public Messaging::UnsolicitedMessageHandler
+{
+
+public:
+    CHIP_ERROR Init(Messaging::ExchangeManager * exchangeManager, ICDClientStorage * clientStorage, CheckInDelegate * delegate);
+    void Shutdown();
+
+    CheckInHandler();
+
+    virtual ~CheckInHandler() = default;
+
+    /**
+     * @brief Callback received on successfully establishing a CASE session in order to re-register the client with the peer node
+     * using a new key to avoid counter rollover problems.
+     *
+     * @param[in] context context of the client establishing the CASE session
+     * @param[in] exchangeMgr exchange manager to use for the re-registration
+     * @param[in] sessionHandle session handle to use for the re-registration
+     */
+    static void HandleDeviceConnected(void * context, Messaging::ExchangeManager & exchangeMgr,
+                                      const SessionHandle & sessionHandle);
+    /**
+     * @brief Callback received on failure to establish a CASE session in order to re-register the client with the peer node using a
+     * new key to avoid counter rollover problems.
+     *
+     * @param[in] context context of the client establishing the CASE session
+     * @param[in] peerId Scoped Node ID of the peer node
+     * @param[in] err failure reason
+     */
+    static void HandleDeviceConnectionFailure(void * context, const ScopedNodeId & peerId, CHIP_ERROR err);
+
+    class RegisterCommandSenderCallback : public CommandSender::Callback
+    {
+    public:
+        void OnResponse(chip::app::CommandSender * apCommandSender, const chip::app::ConcreteCommandPath & aPath,
+                        const chip::app::StatusIB & aStatus, chip::TLV::TLVReader * aData) override
+        {}
+        void OnError(const chip::app::CommandSender * apCommandSender, CHIP_ERROR aError) override { mError = aError; }
+        void OnDone(chip::app::CommandSender * apCommandSender) override {}
+
+        CHIP_ERROR mError = CHIP_NO_ERROR;
+    } registerCommandSenderDelegate;
+
+protected:
+    // ExchangeDelegate
+    CHIP_ERROR
+    OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                      System::PacketBufferHandle && payload) override;
+
+    // UnsolicitedMessageHandler
+    CHIP_ERROR OnUnsolicitedMessageReceived(const PayloadHeader & payloadHeader, ExchangeDelegate *& newDelegate) override;
+
+    // TODO : Follow up to check if this really needs to be a pure virtual function in Exchange delegate
```

Comment:
Please make sure the issue is filed and reference it here.  No one will see this TODO comment.

Reply:
https://github.com/project-chip/connectedhomeip/issues/31322